src/osd/PeeringState.cc

```diff
@@ -1217,120 +1209,129 @@ void PeeringState::reject_reservation()
-      max_last_epoch_started_found = i->second.history.last_epoch_started;
+      max_les = info.history.last_epoch_started;
     }
-    if (!i->second.is_incomplete() &&
-	max_last_epoch_started_found < i->second.last_epoch_started) {
+    if (!info.is_incomplete() &&
+        max_les < info.last_epoch_started) {
       *history_les_bound = false;
-      max_last_epoch_started_found = i->second.last_epoch_started;
+      max_les = info.last_epoch_started;
     }
   }
-  for (map<pg_shard_t, pg_info_t>::const_iterator i = infos.begin();
-       i != infos.end();
-       ++i) {
-    if (max_last_epoch_started_found <= i->second.last_epoch_started) {
-      if (min_last_update_acceptable > i->second.last_update)
-	min_last_update_acceptable = i->second.last_update;
+  for ([[maybe_unused]] auto& [shard, info] : infos) {
+    if (max_les <= info.last_epoch_started &&
+        min_last_update_acceptable > info.last_update) {
+      min_last_update_acceptable = info.last_update;
     }
   }
-  if (min_last_update_acceptable == eversion_t::max())
-    return infos.end();
-
-  map<pg_shard_t, pg_info_t>::const_iterator best = infos.end();
+  if (min_last_update_acceptable == eversion_t::max()) {
+    return std::nullopt;
+  }
   // find osd with newest last_update (oldest for ec_pool).
   // if there are multiples, prefer
   //  - a longer tail, if it brings another peer into log contiguity
   //  - the current primary
-  for (map<pg_shard_t, pg_info_t>::const_iterator p = infos.begin();
-       p != infos.end();
-       ++p) {
-    if (restrict_to_up_acting && !is_up(p->first) &&
-	!is_acting(p->first))
-      continue;
-    // Only consider peers with last_update >= min_last_update_acceptable
-    if (p->second.last_update < min_last_update_acceptable)
-      continue;
-    // Disqualify anyone with a too old last_epoch_started
-    if (p->second.last_epoch_started < max_last_epoch_started_found)
-      continue;
-    // Disqualify anyone who is incomplete (not fully backfilled)
-    if (p->second.is_incomplete())
-      continue;
-    if (best == infos.end()) {
-      best = p;
-      continue;
+  using candidate_t = typename decay_t<decltype(infos)>::value_type;
+  struct is_good {
+    // std::filterd() requires the predicate to be assignable and copy
+    // constructible. so do it the hard way
+    bool restrict_to_up_acting;
+    eversion_t min_last_update_acceptable;
+    epoch_t max_les;
+    const PeeringState* thiz;
+    is_good(bool restrict_to_up_acting, eversion_t min_lua, epoch_t max_les,
+	    const PeeringState* thiz)
+      : restrict_to_up_acting{restrict_to_up_acting},
+	min_last_update_acceptable{min_lua}, max_les{max_les}, thiz{thiz} {}
+    is_good(const is_good& rhs) = default;
+    is_good& operator=(const is_good& rhs) = default;
+    bool operator()(const candidate_t& pi) const {
+      auto& [shard, info] = pi;
+      if (restrict_to_up_acting &&
+	  !thiz->is_up(shard) && !thiz->is_acting(shard)) {
```

Comment:
PeeringState * is ps elsewhere in the file iirc.

Reply:
updated.